libraries/ESP8266WiFi/src/ArduinoWiFiServer.h

```diff
@@ -0,0 +1,117 @@
+/*
+  ArduinoWiFiServer.h - Arduino compatible WiFiServer
+  implementation for ESP8266Wifi library.
+  Copyright (c) 2020 Juraj Andrassy
+
+  This library is free software; you can redistribute it and/or
+  modify it under the terms of the GNU Lesser General Public
+  License as published by the Free Software Foundation; either
+  version 2.1 of the License, or (at your option) any later version.
+
+  This library is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+  Lesser General Public License for more details.
+
+  You should have received a copy of the GNU Lesser General Public
+  License along with this library; if not, write to the Free Software
+  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+*/
+
+#ifndef arduinowifiserver_h
+#define arduinowifiserver_h
+
+#include <ESP8266WiFi.h>
+
+#ifndef MAX_MONITORED_CLIENTS
+#define MAX_MONITORED_CLIENTS 5
+#endif
+
+template <class TServer, class TClient>
+class ArduinoComptibleWiFiServerTemplate : public TServer {
+public:
+
+  ArduinoComptibleWiFiServerTemplate(const IPAddress& addr, uint16_t port) : TServer(addr, port) {}
+  ArduinoComptibleWiFiServerTemplate(uint16_t port) : TServer(port) {}
+  virtual ~ArduinoComptibleWiFiServerTemplate() {}
+
+  // https://www.arduino.cc/en/Reference/EthernetServerAccept
+  TClient accept() {
+    return TServer::available();
+  }
+
+  // https://www.arduino.cc/en/Reference/WiFiServerAvailable
+  TClient available() {
+
+    // update connected clients
+    for (uint8_t i = 0; i < MAX_MONITORED_CLIENTS; i++) {
+      if (!connectedClients[i]) {
+        connectedClients[i] = accept();
+      }
+    }
+    // find next client with data available
+    for (uint8_t i = 0; i < MAX_MONITORED_CLIENTS; i++) {
+      if (index == MAX_MONITORED_CLIENTS) {
+        index = 0;
+      }
+      TClient& client = connectedClients[index];
+      index++;
+      if (client.available())
+        return client;
+    }
+    return TClient(); // no client with data found
+  }
+
+  virtual size_t write(uint8_t b) override {
+    return write(&b, 1);
+  }
+
+  virtual size_t write(const uint8_t *buf, size_t size) override {
+    size_t ret = 0;
+    for (uint8_t i = 0; i < MAX_MONITORED_CLIENTS; i++) {
```

Comment:
Should `availableForWrite()` be tested for all writable clients and be compared to `size` and use the min value instead of `size` ?

Reply:
yes. to avoid blocking. sorry I am not used to reliable availableForWrite. last time a checked it was missing  in Print here. I usually write code for many platforms. I am little confused with this nice limited space :-). but we can't return how many bytes were written to individual clients. so it is shot and forget

Reply:
> sorry I am not used to reliable availableForWrite

I guess it is reliable in this core :)

> last time a checked it was missing in Print here

[It is there now](https://github.com/esp8266/Arduino/pull/7658), in master, as you know.

> EDIT: I think the user would prefer a little blocking over incomplete data

That's indeed the purpose of using it

> but we can't return how many bytes were written to individual clients. so it is shot and forget

We can get max availableForWrite over all opened client, then write this max size and return it.
The writer would have to deal with the returned value.

Reply:
ok, I code it that way

Reply:
done, but I had the PR7680 merged in the ArduinoWiFiServer branch, now I am not sure of the result. I think I can restore the branch to clean state without the PR7680 if necessary